### PR TITLE
Improvements on the Reindex API to support more options

### PR DESCRIPTION
**Pull request by damienalexandre**
This PR replace #1494 and add some missing options and tests. Thanks to @daviscaics for the initial contribution.

The Reindex API now supports all the options listed on the documentation https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html - tests are added too for Script, remote, etc.

## BC Break

I chose to introduce some BC Breaks and I hope that's ok.

- the `run` method was also calling `refresh()` on the new Index. So it's impossible to reindex an Index and make sure it's not yet refreshed _for some reason, like doing other indexation in that index after the reindex_. It does not make sense to force a refresh on the users. If that's really an issue, I can use the REFRESH Param value with a default value to "true", to keep the current behavior, but still allow to disable it;
- the `run` method was returning the new Index object, that's not consistent with other API and that does not allow to read the response content (is there any conflict, how long does it took...). So I changed the returned value.

## Still not supported

- the ability to set multiple index as the source index

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,14 +23,15 @@ All notable changes to this project will be documented in this file based on the
...
+* Elastica\Reindex missing options (script, remote, wait_for_completion, scroll...)
 
 ### Improvements
 * Added `native_function_invocation` CS rule [#1606](https://github.com/ruflin/Elastica/pull/1606)
 * Elasticsearch test version changed from 6.5.2 to 6.6.1 [#1620](https://github.com/ruflin/Elastica/pull/1620)
 
 ### Deprecated
 
-## [Unreleased](https://github.com/ruflin/Elastica/compare/6.1.0...6.1.1)
+## [6.1.1](https://github.com/ruflin/Elastica/compare/6.1.0...6.1.1)
```
Comment (ruflin) [inline CHANGELOG.md]: Thanks for the fix.

File: lib/Elastica/Reindex.php [hunk 2]
```diff
@@ -32,84 +46,94 @@ class Reindex
...
+    private function _resolveBodyOptions(array $params)
     {
-        return \array_intersect_key($options, [
+        return \array_intersect_key($params, [
             self::SIZE => null,
             self::CONFLICTS => null,
         ]);
     }
 
-    /**
```
Comment (ruflin) [inline lib/Elastica/Reindex.php]: Why did you remove these comments?

File: lib/Elastica/Reindex.php [hunk 3]
```diff
@@ -32,84 +46,94 @@ class Reindex
...
-        $body = $this->_getBody($this->_oldIndex, $this->_newIndex, $this->_options);
+        $body = $this->_getBody($this->_oldIndex, $this->_newIndex, $this->getParams());
 
         $reindexEndpoint = new \Elasticsearch\Endpoints\Reindex();
+        $params = \array_intersect_key($this->getParams(), \array_fill_keys($reindexEndpoint->getParamWhitelist(), null));
+        $reindexEndpoint->setParams($params);
         $reindexEndpoint->setBody($body);
 
-        $this->_oldIndex->getClient()->requestEndpoint($reindexEndpoint);
-        $this->_newIndex->refresh();
```
Comment (ruflin) [inline lib/Elastica/Reindex.php]: Good idea to remove it as I agree we should not enforce the refresh.

File: lib/Elastica/Reindex.php [hunk 4]
```diff
@@ -32,84 +46,94 @@ class Reindex
...
+        $this->_lastResponse = $this->_oldIndex->getClient()->requestEndpoint($reindexEndpoint);
 
-        return $this->_newIndex;
+        return $this->_lastResponse;
```
Comment (ruflin) [inline lib/Elastica/Reindex.php]: +1 on returning the response. The _newIndex will still be accessible through the object but the response would be lost. Should we even store it in the object?

Comment (damienalexandre) [reply lib/Elastica/Reindex.php]: Storing the Response is only useful for `getTaskId()`, which is not that important to me as it's easy to get now (thanks to the response being returned). I will remove this method.

File: lib/Elastica/Reindex.php (hunk 2 above)
Comment (damienalexandre) [reply lib/Elastica/Reindex.php]: To make the code more homogeneous and less noisy - lots of methods are self-documented via type-hint, this PHP DocBlock don't provide any useful information and is on a private method.

Should I put it back?

File: lib/Elastica/Reindex.php (hunk 4 above)
Comment (ruflin) [reply lib/Elastica/Reindex.php]: I missed that you use it for getTaskId. I must confess there could be some convenience for users that know less about reindex to have the getTaskId method available.

File: lib/Elastica/Reindex.php (hunk 2 above)
Comment (ruflin) [reply lib/Elastica/Reindex.php]: How will doc blocks generated by phpdoc look like without the docs? Will an IDE know the type that is returned? If the behaviour with or without stays the same, happy to remove it.

Comment (damienalexandre) [reply lib/Elastica/Reindex.php]: Those are private methods so I don't think it changes anything for the end user.

Also behavior is really good on IDE like phpStorm too. I'm even going one step further by adding return type hint as Elastica is 7.0 :+1:

**Comment by damienalexandre** [comment]
I think we can agree to keep the `_lastReponse` property, as it's convenient for the end-user.

Let me know if there is anything else :pray: :+1:

**Comment by ruflin** [comment]
@damienalexandre Thanks for pushing this over the line.
